src/protocol/sort/generate_permutation.rs

```diff
@@ -220,6 +189,51 @@ where
     Ok(composed_less_significant_bits_permutation)
 }
 
+#[embed_doc_image("semi_honest_sort", "images/sort/semi-honest-sort.png")]
+/// This is an implementation of `GenPerm` (Algorithm 6) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+/// This protocol generates permutation of a stable sort for the given shares of inputs.
+///
+/// Steps
+/// For the 0th bit
+/// 1. Get replicated shares in Field using modulus conversion
+/// 2. Compute bit permutation that sorts 0th bit
+/// For 1st to N-1th bit of input share
+/// 1. Shuffle and reveal the i-1th composition
+/// 2. Get replicated shares in Field using modulus conversion
+/// 3. Sort ith bit based on i-1th bits by applying i-1th composition on ith bit
+/// 4  Compute bit permutation that sorts ith bit
+/// 5. Compute ith composition by composing i-1th composition on ith permutation
+/// In the end, n-1th composition is returned. This is the permutation which sorts the inputs
+///
+/// ![Generate sort permutation steps][semi_honest_sort]
+pub async fn generate_permutation<F>(
+    ctx: SemiHonestContext<'_, F>,
+    sort_keys: &[Vec<Replicated<F>>],
+    num_bits: u32,
+
+    // I am not a big fan of using flags to indicate the branch it needs to take.
+    // I would prefer to have two different functions that perform malicious/semi-honest sort
+    malicious_setting: bool,
+) -> Result<Vec<Replicated<F>>, Error>
+where
+    F: Field,
+{
+    if malicious_setting {
+        let malicious_validator = Some(MaliciousValidator::new(
+            ctx.narrow(&MaliciousUpgradeContext),
+        ));
+        let ctx = malicious_validator.unwrap().context();
```

Comment:
Why not:

```suggestion
        let malicious_validator = MaliciousValidator::new(
            ctx.narrow(&MaliciousUpgradeContext),
        );
        let ctx = malicious_validator.context();
```

You added an `Option<T>` wrapper and then unwrapped it immediately.

Reply:
We were trying to combine the two sorts i.e. malicious and semi honest in the same function hence introduced an option for malicious_validator. Will revisit later